Ensure assertion value is double quoted

molovo opened the pull request:
Internally, the `assert` helper does not quote the value to be asserted
against when cascading the call to the relevant assertion function,
resulting in an implicit pass if the value is empty, since the
comparisons are shifted forward a step. Double quoting the vaue when
cascading prevents these false-positives from occurring.

### Todo

- [ ] Write tests to capture these false-positives

Fix #84

molovo committed:
Ensure assertion value is double quoted

Internally, the `assert` helper does not quote the value to be asserted
against when cascading the call to the relevant assertion function,
resulting in an implicit pass if the value is empty, since the
comparisons are shifted forward a step. Double quoting the vaue when
cascading prevents these false-positives from occurring.

- [ ] Write tests to capture these false-positives

Fix #84

MichaelAquilina commented on src/helpers.zsh:
```diff
@@ -144,7 +144,7 @@ function assert() {
 
   # Run the assertion
-  "_zunit_assert_${assertion}" $value ${(@f)comparisons[@]}
+  "_zunit_assert_${assertion}" "$value" ${(@f)comparisons[@]}
```
Is this what is causing `contains` to behave strangely?

molovo commented:
Yeah it is. It's not just contains it actually affected all assertions. If the value was empty the missing quotes meant the arguments passed to the assertion were all shifted left

MichaelAquilina commented:
Good catch @molovo 👍

molovo commented:
Although this does fix the issue, I'm having trouble writing tests to catch the false positives, since the empty value isn't being passed to `assert` when called by the `run` helper either. I expect it's a similiar issue but it needs fixing before this can be merged.

molovo committed:
Quote command arguments passed to run helper

Prevents argument lists from being shifted left if one of the arguments
is an empty string

molovo commented:
@MichaelAquilina How do you think empty strings should be handled with the `equals`, `does_not_equal`, `is_less_than` and `is_greater_than` assertions? At the moment `assert '' equals 0` passes, since the numeric equivalent of an empty string is 0. Technically this is correct, but might result in false-positives again when passing an empty variable. Do you think it's worth adding a check to ensure that only numeric values can be passed to those assertions, since we have the `same_as` and `different_to` assertions for strings anyway?

molovo committed:
Add tests for assertion false-positives with empty values

MichaelAquilina commented:
I think considering that this is a testing framework it would be smart to check the types being passed in for each method. We dont specify that `equals` is equivalent to `-eq` in bash anywhere in the docs so we dont have to follow the same rules. So I would say go ahead with type checking for all of them :)

MichaelAquilina commented on tests/assertions/does_not_contain.zunit:
```diff
@@ -7,6 +7,13 @@
 }
 
+@test 'Test _zunit_assert_does_not_contain success with empty value' {
+    run assert '' does_not_contain 'foo'
```
we could probably kill a lot of repeated functionality with a `assert_not` function

molovo commented:
Going to merge this now and revisit [this idea](https://github.com/zunit-zsh/zunit/pull/90?_pjax=%23js-repo-pjax-container#issuecomment-327470907) for an `assert_not` method in `v1.0.0`
